articles/151_roslaunch_xml.md

```diff
@@ -0,0 +1,234 @@
+
+##### Examples
+
+```xml
+<include file="/opt/my_launch_file.py" ns="/my_ns"/>
+<include file="$(find-pkg my_pkg)/launch/some_launch_file.xml"/>
+<include file="/opt/my_other_launch_file.xml">
+  <arg name="some_argument" value="dummy_value"/>
+</include>
+```
+
+#### `<group>` Tag
+
+##### Description
+
+The `<group>` tag allows for launch actions' grouping as well as optional launch file configuration scoping.
+
+##### Examples
+
+```xml
+<group ns="dummy_group" scoped="true">
+  <node package="a_ros_package" name="dummy0" executable="dummy_node"/>
+  <node package="a_ros_package" name="dummy1" executable="dummy_node"/>
+</group>
+```
+
+Namespaces for groups are usually relative to allow further namespacing by other launch files including the one the group is in.
+
+#### `<let>` Tag
+
+##### Description
+
+The `<let>` tags allows for definition of scoped launch file configuration variables.
+
+##### Examples
+
+```
+<let var="foo" value="$(env BAR)"/>
+<let var="baz" value="false"/>
+```
+
+#### `<arg>` Tag
+
+##### Description
+
+The `<arg>` tag allows for launch file configuration via the command-line or when including it via an `<include>` tag.
+Arguments are launch configuration variables just like the ones defined by `<let>` tags.
+Arguments are limited to the scope of their definition and thus have to be explictly passed to included files if any.
+
+##### Examples
+
+```xml
+<arg name="publish_frequency" default="10"/>
+<arg name="output_path" description="Output path for some processing pipeline"/>
+```
+
+#### `<node>` Tag
+
+##### Description
+
+The `<node>` tag allows for executing a ROS node in the form of local OS process.
+
+##### Examples
+
+```xml
+<node package="ros_demos" executable="publisher">
+  <param name="publish_frequency" value="10"/>
+  <remap from="generic_topic_name" to="my_topic"/>
+</node>
+```
+
+#### `<param>` Tag
+
+##### Description
+
+The `<param>` tag allows for setting a ROS parameter of a ROS node.
+
+##### Examples
+
+```xml
+<node package="my_pkg" executable="my_node">
+  <param name="some_numeric_param" value="100.2"/>
+  <param name="some_list_param" value="Some phrase,100.0,true" sep=","/>
+</node>
+```
+
+#### `<params>` Tag
+
+##### Description
+
+The `<params>` tag allows to either bring ROS parameters from a YAML parameters file or to nest a `<param>` definitions under an appropriate name (i.e. to make a map of them).
+
+##### Examples
+
+```xml
+<node package="my_pkg" executable="my_node">
+  <params from="path/to/param/file.yml"/>
+  <params ns="some_param_group">
+    <param name="some_integer_param" value="10"/>
+  </params>
+</node>
+```
+
+#### `<remap>` Tag
+
+##### Description
+
+The `<remap>` tag allows for ROS name remapping of a `<node>` instance.
+
+##### Examples
+
+```xml
+<node package="my_pkg" executable="my_node">
+  <remap from="chatter" to="/my_chatter"/>
+  <remap from="*/stuff" to="private_\1/stuff"/>
```

Comment:
The globing scope of this is presumably all ros topics sub'ed or pub'ed by `my_node` that end with `/stuff` sould be remapped to a single topic?

Reply:
Following the [name remapping design document](http://design.ros2.org/articles/static_remapping.html), above's rule means that all topics sub'ed and/or pub'ed by `my_node` whose name matches `([^/]+)/stuff` would be mapped to `private_\1/stuff` where `\1` is the first capturing group in the matching pattern.

Reply:
That's a nice idea. And the launch system is intended to be introspectable. But I'm not sure how much of the remapping functionality is exposed to perform such dry runs.

Reply:
You could introspect launch without doing anything, extracting all information you might want, and then use `rclpy`'s functions to expand them "off-line". It should be possible to at least get the exact (and fully expanded) remapping arguments that would be passed to nodes without actually running them.

However, in the specific case you're discussing, it's impossible to know what topics might be created during runtime which may be affected by this, if that's the information you're after.

In order to do that, you'd need a Node IDL which described which topics each node would create if run. This would be easy to implement in most case but is problematic to cover dynamic cases, e.g. move base creates topics based on parameter values (one for each source of obstacle avoidance).